slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,316 @@
+			}
+			pm.handlePodUpdate(p)
+			return nil
+		},
+		func(obj interface{}) error {
+			p, ok := obj.(*v1.Pod)
+			if !ok {
+				if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+					p, ok = d.Obj.(*v1.Pod)
+					if !ok {
+						glog.Errorf("Failed to cast embedded object from tombstone to *v1.Pod: %v", d.Obj)
+						return nil
+					}
+				} else {
+					glog.Errorf("Failed to cast observed object to *v1.Pod: %v", obj)
+					return nil
+				}
+			}
+			pm.handlePodDelete(p)
+			return nil
+		},
+	)
+	go controller.Run(stopCh)
+
+	eventSelector := fields.Set{
+		"involvedObject.kind": "Pod",
+		// this is actually source.Component
+		"source": "kubelet",
+	}.AsSelector().String()
+
+	eventController := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = eventSelector
+				obj, err := pm.kubeClient.Core().Events(NamespaceSystem).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = eventSelector
+				return pm.kubeClient.Core().Events(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Event{},
+		func(obj interface{}) error {
+			e, ok := obj.(*v1.Event)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Event: %v", obj)
+				return nil
+			}
+			pm.handleEventUpdate(e)
+			return nil
+		},
+		func(obj interface{}) error {
+			return nil
+		},
+	)
+	go eventController.Run(stopCh)
+	wait.Until(pm.purgeOutdated, 15*time.Second, stopCh)
+	return nil
+}
+
+func (pm *PodStartupLatencyDataMonitor) purgeOutdated() {
+	now := time.Now()
+	pm.Lock()
+	defer pm.Unlock()
+	if len(pm.toDelete) == 0 {
+		return
+	}
+	left := []string{}
+	for _, podKey := range pm.toDelete {
+		if data, ok := pm.PodStartupData[podKey]; ok {
+			if data.deletedTimestamp.Add(pm.purgeAfter).After(now) {
+				left = append(left, podKey)
+				continue
+			}
+			delete(pm.PodStartupData, podKey)
+		}
+	}
+
+	pm.toDelete = left
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodUpdate(p *v1.Pod) {
+	if isReady, create, running := checkPodAndGetStartupLatency(p); isReady {
+		go pm.insertPodRunningTime(getPodKey(p), create, running, len(p.Spec.Containers))
+	}
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePodDelete(p *v1.Pod) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	key := getPodKey(p)
+	data, ok := pm.PodStartupData[key]
+	if !ok {
+		data = PodStartupMilestones{}
+		data.startedPulling = time.Unix(0, 0)
+	}
+	data.deletedTimestamp = time.Now()
+	pm.PodStartupData[key] = data
+	pm.toDelete = append(pm.toDelete, key)
+}
+
+func (pm *PodStartupLatencyDataMonitor) handlePullingImageEvent(key string, e *v1.Event) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	ok := false
+	data := PodStartupMilestones{}
+	if data, ok = pm.PodStartupData[key]; !ok {
+		data.finishedPulling = time.Unix(0, 0)
+		data.needPulled = -1
+	}
+	data.startedPulling = e.FirstTimestamp.Time
```

Comment:
Hmm - I think this is not correct. Basically, if we have multiple images to pull, shouldn't we use the smallest one here?